ibis/expr/signature.py

```diff
@@ -112,37 +106,30 @@ def Argument(validator, default=EMPTY):
 
 class AnnotableMeta(type):
     def __new__(metacls, clsname, bases, dct):
-        slots, params = [], {}
-
+        params = {}
         for parent in bases:
-            # inherit parent slots
-            if hasattr(parent, '__slots__'):
-                slots += parent.__slots__
             # inherit from parent signatures
             if hasattr(parent, '__signature__'):
                 params.update(parent.__signature__.parameters)
 
+        slots = list(dct.pop('__slots__', []))
         attribs = {}
         for name, attrib in dct.items():
             if isinstance(attrib, Validator):
                 # so we can set directly
                 params[name] = Parameter(name, validator=attrib)
+                slots.append(name)
             else:
                 attribs[name] = attrib
 
-        # if slots or signature are defined no inheritance happens
-        slots = attribs.get('__slots__', tuple(slots))
-        slots += tuple(params.keys())
-
         # mandatory fields without default values must preceed the optional
         # ones in the function signature, the partial ordering will be kept
         params = sorted(
             params.values(), key=lambda p: p.default is EMPTY, reverse=True
         )
-        signature = attribs.get('__signature__', inspect.Signature(params))
```

Comment:
Haven not used this feature anywhere.

Reply:
This is fine to remove. We could potentially use it in a world where we use `matchpy` to implement algebraic optimizations, but that's a ways out.